beacon-chain/rpc/eth/helpers/sync.go

```diff
@@ -53,26 +56,106 @@ func ValidateSync(
 	return status.Error(codes.Unavailable, "Syncing to latest head, not ready to respond")
 }
 
-// IsOptimistic checks whether the latest block header of the passed in beacon state is the header of an optimistic block.
-func IsOptimistic(ctx context.Context, st state.BeaconState, optimisticModeFetcher blockchain.OptimisticModeFetcher) (bool, error) {
-	header := st.LatestBlockHeader()
-	// This happens when the block at the state's slot is not missing.
-	if bytes.Equal(header.StateRoot, params.BeaconConfig().ZeroHash[:]) {
-		root, err := st.HashTreeRoot(ctx)
-		if err != nil {
-			return false, errors.Wrap(err, "could not get state root")
+// IsOptimistic checks whether the beacon state's block is optimistic.
+func IsOptimistic(
+	ctx context.Context,
+	stateId []byte,
+	optimisticModeFetcher blockchain.OptimisticModeFetcher,
+	stateFetcher statefetcher.Fetcher,
+	chainInfo blockchain.ChainInfoFetcher,
+	database db.ReadOnlyDatabase,
+) (bool, error) {
+	stateIdString := strings.ToLower(string(stateId))
+	switch stateIdString {
+	case "head":
+		return optimisticModeFetcher.IsOptimistic(ctx)
+	case "genesis", "finalized":
+		return false, nil
```

Comment:
Why wouldn't we check finalized?
```
fcp := chainInfo.FinalizedJustifiedCheckpoint()
		if fcp == nil {
			return true, errors.New("received nil finalized checkpoint")
		}
		return optimisticModeFetcher.IsOptimisticForRoot(ctx, bytesutil.ToBytes32(fcp.Root))
```

Reply:
finalized block can't be optimistic, as said on Slack